pkg/ctrlz/ctrlz.go

```diff
@@ -133,14 +164,50 @@ func Run(o *Options, customTopics []fw.Topic) {
 		addr = ""
 	}
 
-	s := &http.Server{
-		Addr:           fmt.Sprintf("%s:%d", addr, o.Port),
-		Handler:        router,
-		ReadTimeout:    10 * time.Second,
-		WriteTimeout:   10 * time.Second,
-		MaxHeaderBytes: 1 << 20,
+	shutdownMutex.Lock()
+
+	if abort {
+		// reset to default state in case Run is called again
+		abort = false
+	} else {
+		server = &http.Server{
+			Addr:           fmt.Sprintf("%s:%d", addr, o.Port),
+			Handler:        router,
+			ReadTimeout:    10 * time.Second,
+			WriteTimeout:   10 * time.Second,
+			MaxHeaderBytes: 1 << 20,
+		}
+		shutdown.Add(1)
+	}
+
+	shutdownMutex.Unlock()
+
+	if server != nil {
+		log.Infof("ControlZ available at %s:%d", getLocalIP(), o.Port)
+		server.ListenAndServe()
+		log.Infof("ControlZ terminated")
+		shutdown.Done()
+		server = nil
+	}
+}
+
+// Stop terminates ControlZ.
+//
+// Stop is hot normally used by programs that expose ControlZ, it is primarily intended to be
+// used by tests.
+func Stop() {
+	wait := false
+
+	shutdownMutex.Lock()
+	if server != nil {
+		server.Close()
+		wait = true
+	} else {
+		abort = true
```

Comment:
The API is inherently racy in nature, since Run returns no value. 

abort is there is preclude outstanding calls to Run from succeeding. This is necessary, otherwise if you have code that calls Run in one goroutine, and different (test) code calling Stop, you can never be sure that it'll work properly as Stop might end up being called before Run finishes. The abort flag prevents this race and ensures proper termination.